_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+
+```rust
+use libc::size_t;
+
+#[link(name = "snappy")]
+extern {
+    fn snappy_max_compressed_length(source_length: size_t) -> size_t;
+}
+
+fn main() {
+    let x = unsafe { snappy_max_compressed_length(100) };
+    println!("max compressed length of a 100 byte buffer: {}", x);
+}
+```
+
+Note that the call from Rust to C requires the `unsafe` keyword.
+This recognises the fact that Rust enforces a lot of useful properties, such as lifetime safety, that C does not and, as soon as you transition into C code, these properties are not automatically enforced.
+As a Rust programmer, you are expected to enforce these properties at the API boundary, but in the general case that is impossible on most hardware.
+For example, there's no Rust wrapper that you can write that prevents a C library (on non-CHERI hardware) from mutating an immutable object, or from capturing a borrowed reference.
+
+But what if the hardware could help?
+What if the Rust compiler could tell some combination of the hardware and the language-agnostic run-time system that a pointer must not be captured, or may be used with load instructions but not store?
+
+## So what is this chariot thing you kids are talking about?
+
+CHERIoT allows a compiler to enforce a fairly rich set of properties against other bits of a program.
+It can guarantee that, when invoking untrusted code, the untrusted code cannot:
+
+ - Access objects unless passed pointers to them.
+ - Access an object (or the memory that was formerly used for the object) after the object has been freed.
+ - Hold a pointer to an object with automatic storage duration (an 'on-stack' object) after the end of the call in which it was created.
+ - Hold a temporarily delegated pointer beyond a single call.
+ - Modify an object passed via immutable reference.
+ - Modify any object reachable from an object that is passed as a deeply immutable reference.
+ - Tamper with (or access) an object passed via opaque reference.
+
+Some of these are baseline guarantees from CHERI, some are built atop those with additional hardware extensions, and others are built atop them with some language-agnostic software.
+Let's go through them one at a time.
+
+Running code cannot access objects unless passed pointers to them.
+This is a basic property of CHERI but also the absolute minimum foundation for *any* form of memory safety.
+If malicious or buggy code can materialise arbitrary pointers out of thin air and use them, it can bypass any of the other rules.
+In any CHERI system, all instructions that accesses memory (loads, stores, jumps, and so on) must take a capability as one of its operands.
+A capability (in general) is an unforgeable token of authority that allows you to do some action.
+A CHERI capability is a machine word that is interpreted by the CPU and which can be used as a hardware type that *represents the language notion of a pointer*.
```

Comment:
```suggestion
A CHERI capability is a machine word that is interpreted by, and protected by, the CPU and which can be used as a hardware type that *represents the language notion of a pointer*.
```

Reply:
I also think I would prefer 'implements' or even 'reifies' to 'represents'.